### ISSUE TITLE
--rustfmt_bindings doesn't work with printing bindings to stdout

### ISSUE THREAD

**Issue by fitzgen** [rust-lang-nursery/rust-bindgen#968]
This produces pretty code in `output.rs`:

```
$ bindgen input.h --rustfmt-bindings -o output.rs
```

This does *not* print pretty code to stdout:

```
$ bindgen input.h --rustfmt-bindings
```

See the `rustfmt_generated_file` function in `src/lib.rs` for where we currently run `rustfmt` on generated files. We need to have a version of this for when we are printing to `stdout` that writes the unformatted bindings to a child `rustfmt`'s stdin, and then pipes the child's stdout to our stdout. We actually have something pretty similar to this in `tests/tests.rs` as well in the `rustfmt` function there.

**Comment by myromeu** [comment]
Can I pick up this?

**Comment by fitzgen** [comment]
@myromeu all yours! For the record, you can ask the @highfive bot to automatically assign you:

> If you intend to work on this issue, then add `@highfive: assign me` to your comment, and I'll assign this to you. 😄

Let me know if you have an questions or anything!

**Comment by manaskarekar** [comment]
I'm new to Rust and just learning as I go along, thought this was good place to jump in. I know @myromeu is assigned this issue but I just thought I'd play around with it in parallel.

It's incomplete and needs work and I need to write tests for it, but I thought I'd throw this out there and be steered back on course. Any help/comments would be appreciated. This is mostly hacked together from the two places mentioned in the issue description, but it's helping me learn the syntax and idioms.

This is what I came up with. Comments welcome, thank you!

    impl Bindings { 
        //... truncated

        /// Write these bindings as source text to stdout.
        pub fn write_to_stdout(&self) -> io::Result<()> {
            let bindings = self.to_string();
            self.rustfmt_generated_string(bindings)
        }
        
        /// Checks if rustfmt_bindings is set and runs rustfmt on the string
        fn rustfmt_generated_string(&self, source: String) -> io::Result<()> {
            let _t = self.context.timer("rustfmt_generated_string");
    
            if !self.context.options().rustfmt_bindings {
                return Ok(());
            }
    
            let rustfmt = if let Ok(rustfmt) = which::which("rustfmt") {
                rustfmt
            } else {
                return Err(io::Error::new(
                    io::ErrorKind::Other,
                    "Rustfmt activated, but it could not be found in global path.",
                ));
            };
            
            let mut cmd = Command::new(rustfmt);
            cmd
                .stdin(Stdio::piped())
                .stdout(Stdio::piped())
                .stderr(Stdio::piped());
                
            if let Some(path) = self.context
                .options()
                .rustfmt_configuration_file
                .as_ref()
                .and_then(|f| f.to_str())
            {
                cmd.args(&["--config-path", path]);
            }
            
            let mut child = cmd.spawn().expect("Failed to spawn rustfmt.");
            let mut child_stdin = child.stdin.take().unwrap();
            let mut child_stdout = child.stdout.take().unwrap();
            let mut child_stderr = child.stderr.take().unwrap();
        
            // Write to stdin in a new thread, so that we can read from stdout on this
            // thread. This keeps the child from blocking on writing to its stdout which
            // might block us from writing to its stdin.
            let stdin_handle = ::std::thread::spawn(move || {
                child_stdin.write_all(source.as_bytes())
            });
        
            // Read stderr on a new thread for similar reasons.
            let stderr_handle = ::std::thread::spawn(move || {
              let mut output = vec![];
              io::copy(&mut child_stderr, &mut output)
                  .map(|_| String::from_utf8_lossy(&output).to_string())
            });
        
            let mut stdout = io::stdout();
            io::copy(&mut child_stdout, &mut stdout)
                .expect("Should copy child.stdout into stdout OK");
        
            // Ignore actual rustfmt status because it is often non-zero for trivial
            // things.
            let _ = child.wait().expect("should wait on rustfmt child OK");
        
            stdin_handle.join()
                .expect("writer thread should not have panicked")
                .expect("should have written to child rustfmt's stdin OK");
        
            //TODO: Check for utf8 validity
            //let bindings = String::from_utf8(output)
            //  .expect("rustfmt should only emit valid utf-8");
        
            stderr_handle.join()
              .expect("stderr reader thread should not have panicked")
              .expect("should have read child rustfmt's stderr OK");
        
            Ok(())
        }
        //... truncated
    }

**Comment by fitzgen** [comment]
@myromeu making any progress? Anything I can help with?

**Comment by manaskarekar** [comment]
I'm new to Rust and just learning as I go along, thought this was good place to jump in. I know @myromeu is assigned this issue but I just thought I'd play around with it in parallel.

Sorry, I haven't been able to put in a lot of time in to this. It's incomplete and needs all kinds of fixes work and I need to write tests for it, but I thought I'd throw this out there and be steered back on course. 
Any help/comments would be appreciated. This is mostly hacked together from the two places mentioned in the issue description, but it's helping me learn the syntax and idioms.

Help welcome, thank you!

I am reading up on error handling to deal with the return values better. Right now, the code doesn't compile. 
Was also trying to look up where write_to_file (and consequently write_to_string) is called?



    impl Bindings { 
        //... truncated

        /// Write these bindings as source text to stdout.
        pub fn write_to_stdout(&self) -> io::Result<()> {
            let bindings = self.to_string();
            
            //TODO: Inspect stderr and then re-raise?
            //let (rustfmt_bindings, stderr) = self.rustfmt_generated_string(bindings);
            
            let mut rustfmt_bindings = String::new();
            
            match self.rustfmt_generated_string(bindings) {
                Ok((rustfmt_bindings, stderr)) => Ok(()),
                Err(e) => Err(e)
            }
            
            let mut stdout = io::stdout();
            io::copy(&mut rustfmt_bindings, &mut stdout)
                .expect("Should copy child.stdout into stdout OK");
                
            Ok(())
        }
        
        /// Checks if rustfmt_bindings is set and runs rustfmt on the string
        fn rustfmt_generated_string(&self, source: String) -> io::Result<(String, String)> {
            let _t = self.context.timer("rustfmt_generated_string");
    
            if !self.context.options().rustfmt_bindings {
                return Ok(());
            }
    
            let rustfmt = if let Ok(rustfmt) = which::which("rustfmt") {
                rustfmt
            } else {
                return Err(io::Error::new(
                    io::ErrorKind::Other,
                    "Rustfmt activated, but it could not be found in global path.",
                ));
            };
            
            let mut cmd = Command::new(rustfmt);
            cmd
                .args(&["--write-mode=display"])
                .stdin(Stdio::piped())
                .stdout(Stdio::piped())
                .stderr(Stdio::piped());
                
            if let Some(path) = self.context
                .options()
                .rustfmt_configuration_file
                .as_ref()
                .and_then(|f| f.to_str())
            {
                cmd.args(&["--config-path", path]);
            }
            
            let mut child = cmd.spawn().expect("Failed to spawn rustfmt.");
            let mut child_stdin = child.stdin.take().unwrap();
            let mut child_stdout = child.stdout.take().unwrap();
            let mut child_stderr = child.stderr.take().unwrap();
        
            // Write to stdin in a new thread, so that we can read from stdout on this
            // thread. This keeps the child from blocking on writing to its stdout which
            // might block us from writing to its stdin.
            let stdin_handle = ::std::thread::spawn(move || {
                child_stdin.write_all(source.as_bytes())
            });
        
            // Read stderr on a new thread for similar reasons.
            let stderr_handle = ::std::thread::spawn(move || {
              let mut output = vec![];
              io::copy(&mut child_stderr, &mut output)
                  .map(|_| String::from_utf8_lossy(&output).to_string())
            });
    
            let mut output = vec![];
            io::copy(&mut child_stdout, &mut output)
                .expect("Should copy stdout into vec OK");
                
            // Ignore actual rustfmt status because it is often non-zero for trivial
            // things.
            let _ = child.wait().expect("should wait on rustfmt child OK");
        
            stdin_handle.join()
                .expect("writer thread should not have panicked")
                .expect("should have written to child rustfmt's stdin OK");
        
            let bindings = String::from_utf8(output)
              .expect("rustfmt should only emit valid utf-8");
        
            let child_stderr = stderr_handle.join()
              .expect("stderr reader thread should not have panicked")
              .expect("should have read child rustfmt's stderr OK");
        
            (bindings, child_stderr)
        }
        //... truncated
    }

**Comment by fitzgen** [comment]
Fixed in #1042
